File: samples_c/s95_redis_stringmatchlen.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

/* Kaynak: Redis src/util.c, stringmatchlen_impl()/stringmatchlen() (KEYS
   komutunun glob eslestiricisi) - birebir.
   https://raw.githubusercontent.com/redis/redis/unstable/src/util.c
   Lisans: BSD-3-Clause.
   Girdi: iki satir - once desen, sonra metin (buyuk/kucuk harf DUYARLI,
   nocase=0).
   Cikti: "MATCH" veya "NOMATCH". */

static int stringmatchlen_impl(const char *pattern, int patternLen,
        const char *string, int stringLen, int nocase, int *skipLongerMatches, int nesting)
{
    if (nesting > 1000) return 0;

    while(patternLen && stringLen) {
        switch(pattern[0]) {
        case '*':
            while (patternLen && pattern[1] == '*') {
                pattern++;
                patternLen--;
            }
            if (patternLen == 1)
                return 1;
            while(stringLen) {
                if (stringmatchlen_impl(pattern+1, patternLen-1,
                            string, stringLen, nocase, skipLongerMatches, nesting+1))
                    return 1;
                if (*skipLongerMatches)
                    return 0;
                string++;
                stringLen--;
            }
            *skipLongerMatches = 1;
            return 0;
            break;
        case '?':
            string++;
            stringLen--;
            break;
        case '[':
        {
            int not, match;
            pattern++;
            patternLen--;
            not = patternLen && pattern[0] == '^';
            if (not) {
                pattern++;
                patternLen--;
            }
            match = 0;
            while(1) {
                if (patternLen >= 2 && pattern[0] == '\\') {
                    pattern++;
                    patternLen--;
                    if (pattern[0] == string[0])
                        match = 1;
                } else if (patternLen == 0) {
                    pattern--;
                    patternLen++;
                    break;
                } else if (pattern[0] == ']') {
                    break;
                } else if (patternLen >= 3 && pattern[1] == '-') {
                    int start = pattern[0];
                    int end = pattern[2];
                    int c = string[0];
                    if (start > end) {
                        int t = start;
                        start = end;
                        end = t;
                    }
                    if (nocase) {
                        start = tolower(start);
                        end = tolower(end);
                        c = tolower(c);
                    }
                    pattern += 2;
                    patternLen -= 2;
                    if (c >= start && c <= end)
                        match = 1;
                } else {
                    if (!nocase) {
                        if (pattern[0] == string[0])
                            match = 1;
                    } else {
                        if (tolower((int)pattern[0]) == tolower((int)string[0]))
                            match = 1;
                    }
                }
                pattern++;
                patternLen--;
            }
            if (not)
                match = !match;
            if (!match)
                return 0;
            string++;
            stringLen--;
            break;
        }
        case '\\':
            if (patternLen >= 2) {
                pattern++;
                patternLen--;
            }
            /* fall through */
        default:
            if (!nocase) {
                if (pattern[0] != string[0])
                    return 0;
            } else {
                if (tolower((int)pattern[0]) != tolower((int)string[0]))
                    return 0;
            }
            string++;
            stringLen--;
            break;
        }
        pattern++;
        patternLen--;
        if (stringLen == 0) {
            while(patternLen && *pattern == '*') {
                pattern++;
                patternLen--;
            }
            break;
        }
    }
    if (patternLen == 0 && stringLen == 0)
        return 1;
    return 0;
}

int stringmatchlen(const char *pattern, int patternLen,
        const char *string, int stringLen, int nocase) {
    int skipLongerMatches = 0;
    return stringmatchlen_impl(pattern,patternLen,string,stringLen,nocase,&skipLongerMatches,0);
}

int main(void) {
    char pattern[256], text[256];
    if (!fgets(pattern, sizeof(pattern), stdin)) return 0;
    if (!fgets(text, sizeof(text), stdin)) return 0;
    pattern[strcspn(pattern, "\r\n")] = 0;
    text[strcspn(text, "\r\n")] = 0;
    int r = stringmatchlen(pattern, (int)strlen(pattern), text, (int)strlen(text), 0);
    printf("%s\n", r ? "MATCH" : "NOMATCH");
    return 0;
}

File: translations_rust/s95_redis_stringmatchlen.rs
use std::io::{self, BufRead};

fn stringmatchlen_impl(
    pattern: &[u8],
    string: &[u8],
    nocase: bool,
    skip_longer_matches: &mut bool,
    nesting: i32,
) -> bool {
    if nesting > 1000 {
        return false;
    }

    let mut pi = 0usize;
    let mut si = 0usize;
    let plen0 = pattern.len();
    let slen0 = string.len();

    while pi < plen0 && si < slen0 {
        match pattern[pi] {
            b'*' => {
                while pi + 1 < plen0 && pattern[pi + 1] == b'*' {
                    pi += 1;
                }
                if pi == plen0 - 1 {
                    return true;
                }
                while si < slen0 {
                    if stringmatchlen_impl(&pattern[pi + 1..], &string[si..], nocase, skip_longer_matches, nesting + 1) {
                        return true;
                    }
                    if *skip_longer_matches {
                        return false;
                    }
                    si += 1;
                }
                *skip_longer_matches = true;
                return false;
            }
            b'?' => {
                si += 1;
            }
            b'[' => {
                pi += 1;
                let mut not = false;
                if pi < plen0 && pattern[pi] == b'^' {
                    not = true;
                    pi += 1;
                }
                let mut is_match = false;
                loop {
                    if pi + 1 < plen0 && pattern[pi] == b'\\' {
                        pi += 1;
                        if pattern[pi] == string[si] {
                            is_match = true;
                        }
                    } else if pi >= plen0 {
                        pi -= 1;
                        break;
                    } else if pattern[pi] == b']' {
                        break;
                    } else if pi + 2 < plen0 && pattern[pi + 1] == b'-' {
                        let mut start = pattern[pi];
                        let mut end = pattern[pi + 2];
                        let mut c = string[si];
                        if start > end {
                            std::mem::swap(&mut start, &mut end);
                        }
                        if nocase {
                            start = start.to_ascii_lowercase();
                            end = end.to_ascii_lowercase();
                            c = c.to_ascii_lowercase();
                        }
                        pi += 2;
                        if c >= start && c <= end {
                            is_match = true;
                        }
                    } else {
                        if !nocase {
                            if pattern[pi] == string[si] {
                                is_match = true;
                            }
                        } else if pattern[pi].to_ascii_lowercase() == string[si].to_ascii_lowercase() {
                            is_match = true;
                        }
                    }
                    pi += 1;
                }
                if not {
                    is_match = !is_match;
                }
                if !is_match {
                    return false;
                }
                si += 1;
            }
            b'\\' => {
                if pi + 1 < plen0 {
                    pi += 1;
                }
                if !nocase {
                    if pattern[pi] != string[si] {
                        return false;
                    }
                } else if pattern[pi].to_ascii_lowercase() != string[si].to_ascii_lowercase() {
                    return false;
                }
                si += 1;
            }
            _ => {
                if !nocase {
                    if pattern[pi] != string[si] {
                        return false;
                    }
                } else if pattern[pi].to_ascii_lowercase() != string[si].to_ascii_lowercase() {
                    return false;
                }
                si += 1;
            }
        }
        pi += 1;
        if si == slen0 {
            while pi < plen0 && pattern[pi] == b'*' {
                pi += 1;
            }
            break;
        }
    }

    pi == plen0 && si == slen0
}

fn stringmatchlen(pattern: &[u8], string: &[u8], nocase: bool) -> bool {
    let mut skip_longer_matches = false;
    stringmatchlen_impl(pattern, string, nocase, &mut skip_longer_matches, 0)
}

fn main() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines();
    let pattern = lines.next().unwrap().unwrap();
    let text = lines.next().unwrap().unwrap();
    let r = stringmatchlen(pattern.as_bytes(), text.as_bytes(), false);
    println!("{}", if r { "MATCH" } else { "NOMATCH" });
}
